### PR TITLE
usb: avoid k_yield in usb_write when called from ISR

### PR DESCRIPTION
**Pull request by somagriff**
When usb_dc_ep_write fails in usb_write with EAGAIN, the code was logging
and retrying via k_yield. It seems this function can be called both from
the context of a thread or from an ISR. When called from an ISR we cannot
call k_yield, which in our implementation was causing an assertion,
kernel panic and halt.

Only yield back if we're not in an ISR, otherwise exit immediately with
EAGAIN without retry, and let the caller handle it.

Change the LOG to a printk to avoid similar ISR context issues seen
during logging, whilst still maintaining the log print for completeness.

Signed-off-by: Andy Griffin <agriffin@logitech.com>

### CODE REVIEW

**Review by salggarcia — COMMENTED** [review]
Tried this a couple times but can't get it to get to this point (usb write fail), but this does seem safer than what it was doing

**Comment by somagriff** [comment]
I can provoke this fairly reliably by just unplugging and replugging whilst enumerated on my mac.

**Comment by salggarcia** [comment]
Ok I was able to get it without this fix (and thus crash), but now with this fix it does not crash!

**Comment by nickstoughton** [comment]
Please run `gitlint` on such commits. The first line is too long.

File: subsys/usb/usb_device.c [hunk 1]
```diff
@@ -1335,7 +1335,10 @@ int usb_write(uint8_t ep, const uint8_t *data, uint32_t data_len, uint32_t *byte
 	do {
 		ret = usb_dc_ep_write(ep, data, data_len, bytes_ret);
 		if (ret == -EAGAIN) {
-			LOG_WRN("Failed to write endpoint buffer 0x%02x", ep);
+			printk("Failed to write endpoint buffer 0x%02x", ep);
```
Comment (nickstoughton) [inline subsys/usb/usb_device.c:1338]: We should not be using either `printk` nor `LOG_WRN`. Logging is marginally better than `printk`, but there will be an error message even without this (on before, from `usb_dc_ep_write` and one afterwards from `SendCSW`) and it does nothing but clutter up the output. The error will be clearer without this.

**Review by nickstoughton — CHANGES_REQUESTED** [review]

File: subsys/usb/usb_device.c (hunk 1 above)
Comment (somagriff) [reply subsys/usb/usb_device.c]: "We" are not using, this is upstream code being called in multiple contexts. Can't say I agree with logging an error like this as being clutter -- that should be solved by setting log-level, and remember the msc code is not the only caller of this. So yes, changing to printk would change behavior for all, and if leaving it as LOG means we might still assert, then yes we can avoid the situation by not printing here, but be aware that if the concern is for being an ISR context callee from msc.c that there are numerous functions in the path to get here that also LOG.

But wanting to oil the wheels, so have changed it.

Comment (somagriff) [reply subsys/usb/usb_device.c]: Have run gitlint... had checked this in my editor, but accidentally had it set to 73.

**Review by nickstoughton — APPROVED** [review]

**Comment by somagriff** [comment]
Following further discussion, reworked to use the original strategy of `break`ing, whilst also removing the contentious log message completely.

**Review by nickstoughton — CHANGES_REQUESTED** [review]

File: kernel/timer.c [hunk 2]
```diff
@@ -43,7 +43,10 @@ void z_timer_expiration_handler(struct _timeout *t)
 
 	/* invoke timer expiry function */
 	if (timer->expiry_fn != NULL) {
+		/* Unlock for user handler. */
```
[the remark this answers is not in the document]
Comment (somagriff) [reply kernel/timer.c]: OK, rebased to restore this.

**Review by nickstoughton — APPROVED** [review]

**Pull request merged**